src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/CardinalityBasedGreedyJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,291 @@
+
+                    index++;
+                }
+                else if (pop instanceof PhysicalOpBinaryUnion || pop instanceof PhysicalOpMultiwayUnion) {
+                    int sumCard = 0, sumAccess = 0;
+                    final List<FederationMember> fms = new ArrayList<>();
+                    for ( int count = 0; count < plan.numberOfSubPlans(); count++ ) {
+                        sumCard += resps[index].getCardinality();
+
+                        final FederationMember fm = flattenRequestMemPairs.get(index).getMember();
+                        sumAccess += accessNumForReq( resps[index].getCardinality(), fm );
+                        fms.add( fm );
+
+                        index++;
+                    }
+
+                    planWithStatistics.addCandidate( sumCard );
+                    planWithStatistics.addNumOfAccess( sumAccess );
+                    planWithStatistics.addFederationMembers( fms );
+                }
+                else
+                    throw new IllegalArgumentException("Unsupported type of subquery in source assignment (" + pop.getClass().getName() + ")");
+
+                subPlansWithStatistics.add( planWithStatistics );
+            }
+
+            return subPlansWithStatistics;
+        }
+
+        /**
+         * Compares all available subplans in terms of their cardinality estimation
+         * and returns the one with the lowest estimated cardinality.
+         */
+        protected PhysicalPlanWithStatistics chooseNextSubPlanBasedOnCard( final List<PhysicalPlanWithStatistics> subPlansWithStatistics ) {
+            PhysicalPlanWithStatistics bestPlanWithCandidate = subPlansWithStatistics.get(0);
+            int lowestCost = bestPlanWithCandidate.getCandidate();
+
+            for ( int i = 1; i < subPlansWithStatistics.size(); i++ ) {
+                PhysicalPlanWithStatistics current = subPlansWithStatistics.get(i);
+                if ( current.getCandidate() < lowestCost ) {
+                    lowestCost = current.getCandidate();
+                    bestPlanWithCandidate = current;
+                }
+            }
+
+            return bestPlanWithCandidate;
+        }
+
+        /**
+         * The physical algorithm is determined based on the estimated number of requests to execute the join
+         * - SHJ: Card(T1)/PageSize + Card(T2)/PageSize
+         * - BJ: Card(T1)/PageSize + Card(T1)/blockSize
+         * For the comparison, only keeping the second element is enough.
+         */
+        protected PhysicalPlanWithStatistics decidePhysicalAlgorithm( final PhysicalPlanWithStatistics currentPlan, final PhysicalPlanWithStatistics nextPlan ) {
+            final double accNumSHJ = nextPlan.getNumOfAccess();
+
+            final List<FederationMember> fms = nextPlan.getFederationMembers();
+            double accNumBJ = 0;
+            for ( final FederationMember fm: fms ) {
+                double blockSize = setBlockSize(fm);
```

Comment:
The function name is misleading. You are not actually setting anything here. Better name the function `determineBlockSize`.

Additionally, `blockSize` can be `final`.